pennylane/interfaces/batch/torch.py

```diff
@@ -0,0 +1,236 @@
+    ``tapes``; this Function should always be called
+    with the parameters extracted directly from the tapes as follows:
+
+    >>> parameters = []
+    >>> [parameters.extend(t.get_parameters()) for t in tapes]
+    >>> kwargs = {"tapes": tapes, "device": device, "gradient_fn": gradient_fn, ...}
+    >>> ExecuteTapes.apply(kwargs, *parameters)
+
+    The private argument ``_n`` is used to track nesting of derivatives, for example
+    if the nth-order derivative is requested. Do not set this argument unless you
+    understand the consequences!
+    """
+
+    @staticmethod
+    def forward(ctx, kwargs, *parameters):  # pylint: disable=arguments-differ
+        """Implements the forward pass batch tape evaluation."""
+        ctx.tapes = kwargs["tapes"]
+        ctx.device = kwargs["device"]
+
+        ctx.execute_fn = kwargs["execute_fn"]
+        ctx.gradient_fn = kwargs["gradient_fn"]
+
+        ctx.gradient_kwargs = kwargs["gradient_kwargs"]
+        ctx.max_diff = kwargs["max_diff"]
+        ctx._n = kwargs.get("_n", 1)
+
+        with qml.tape.Unwrap(*ctx.tapes):
+            res, ctx.jacs = ctx.execute_fn(ctx.tapes, **ctx.gradient_kwargs)
+
+        # if any input tensor uses the GPU, the output should as well
+        ctx.torch_device = None
+
+        for p in parameters:
+            if isinstance(p, torch.Tensor):
+                if p.is_cuda:  # pragma: no cover
+                    ctx.torch_device = p.get_device()
+                    break
+
+        for i, r in enumerate(res):
+            if r.dtype == np.dtype("object"):
+                # For backwards compatibility, we flatten ragged tape outputs
+                r = np.hstack(r)
+
+            res[i] = torch.as_tensor(torch.from_numpy(r), device=ctx.torch_device)
+
+            if ctx.jacs:
+                ctx.jacs[i] = torch.as_tensor(
+                    torch.from_numpy(ctx.jacs[i]), device=ctx.torch_device
+                )
+
+        return tuple(res)
+
+    @staticmethod
+    def backward(ctx, *dy):
+        """Returns the vector-Jacobian product with given
+        parameter values p and output gradient dy"""
+
+        if ctx.jacs:
+            # Jacobians were computed on the forward pass (mode="forward")
+            # No additional quantum evaluations needed; simply compute the VJPs directly.
+            vjps = _compute_vjp(dy, ctx.jacs)
+
+        else:
+            # Need to compute the Jacobians on the backward pass (accumulation="backward")
+
+            # Temporary: check if the gradient function is a differentiable transform.
+            # For the moment, simply check if it is part of the `qml.gradients` package.
+            # Longer term, we should have a way of checking this directly
+            # (e.g., isinstance(gradient_fn, GradientTransform))
+            module_name = getattr(inspect.getmodule(ctx.gradient_fn), "__name__", "")
+
+            if "pennylane.gradients" in module_name:

```

Comment:
Yes, I was tempted as well to factor it out, but resisted due to premature optimization 😆 

Since this is the first PR introducing it, I would rather leave it as is, and wait for stability to be reached before making a decision here?

I worry factoring it out may add design complications to complicated bug fixes that might be needed.

Reply:
Sounds good :)